File: 0227.BasicCalculatorII.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = []
        num = 0
        preSign = '+'
        for i in xrange(len(s)):
            if s[i].isdigit():
                num = 10*num + int(s[i])
            if s[i] in '+-*/' or i == len(s)-1:
                if preSign == '+':
                    stack.append(num)
                elif preSign == '-':
                    stack.append(-num)
                elif preSign == '*':
                    stack.append(stack.pop()*num)
                elif preSign == '/' and num != 0:
                    stack.append(int(stack.pop()/float(num)))
                num = 0
                preSign = s[i]
        return sum(stack)

File: 0227.BasicCalculatorII.go
func calculate(s string) int {
    n := len(s)
    stack := []int{}
    num, sign := 0, '+'
    for i, ch := range s {
        if ch >= '0' && ch <= '9' {
            num = 10*num + int(ch-'0')
        }
        if ch=='+' || ch=='-' || ch=='*' || ch=='/' || i==n-1 {
            switch sign {
                case '+':
                stack = append(stack, num)
                case '-':
                stack = append(stack, -num)
                case '*':
                stack[len(stack)-1] *= num
                case '/':
                stack[len(stack)-1] /= num
            }
            num, sign = 0, ch
        }
    }
    ans := 0
    for _, num := range stack {
        ans += num
    }
    return ans
}